### PR TITLE
Polish backward.py to prune more ops

### PR DESCRIPTION
**Pull request by sneaxiy**
This PR tries to fix some issues that users have to set `var.stop_gradient = True` when appending backward. The idea is simple: if the all gradient inputs or outputs of the backward op does not exist, prune this op directly. But there are 2 corner cases that should be concerned:

- the while grad op may take a var that does not exist in forward/parent block as its input. This input can not be found when calling `_append_backward_vars_`. These kinds of vars are the recursive outputs of while grad op, so that it is also the output of while grad op. This PR adds the restriction on this case.
- **`rnn_memory_helper_grad` op takes a var that does not exist in the program as input**. This should not happen in normal ops! But the design of this op is a bit ugly. So I skip this case in this PR, otherwise this op would be pruned, and the calculation result would be wrong.

CE has passed: http://ce.paddlepaddle.org:8080/viewLog.html?buildId=78889&tab=buildResultsDiv&buildTypeId=PaddleModesl_ModelDebug_Model11Build

### CODE REVIEW

File: python/paddle/fluid/backward.py [hunk 1]
```diff
@@ -933,6 +933,31 @@ def _append_backward_ops_(block,
                 cb(block=target_block, context=grad_to_var)
 
 
+def _is_grad_var_(var):
```
Comment (zhhsplendid) [inline python/paddle/fluid/backward.py]: rename "var" as "var_name"?

File: python/paddle/fluid/backward.py [hunk 2]
```diff
@@ -933,6 +933,31 @@ def _append_backward_ops_(block,
...
+    return core.grad_var_suffix() in var
+
+
+# Find the op whose holds the sub_block as its "sub_block" attr
```
Comment (zhhsplendid) [inline python/paddle/fluid/backward.py]: NIT: "Find the op whose holds" -> "Find the op who holds"

**Review by zhhsplendid — COMMENTED** [review]
Other things LGTM

File: python/paddle/fluid/backward.py (hunk 1 above)
Comment (sneaxiy) [reply python/paddle/fluid/backward.py]: Done.

File: python/paddle/fluid/backward.py (hunk 2 above)
Comment (sneaxiy) [reply python/paddle/fluid/backward.py]: Done.

**Review by zhhsplendid — APPROVED** [review]
LGTM now

**Review by Aurelius84 — APPROVED** [review]

**Review by luotao1 — APPROVED** [review]
LGTM for backward.py

**Pull request merged**